### ISSUE TITLE
unified-lint-rule should not set info messages' fatality

### ISSUE THREAD

**Issue by ritave** [remarkjs/remark-lint#298]
### Initial checklist

- [X] I read the [support docs](https://github.com/remarkjs/.github/blob/main/support.md)
- [X] I read the [contributing guide](https://github.com/remarkjs/.github/blob/main/contributing.md)
- [X] I agree to follow the [code of conduct](https://github.com/remarkjs/.github/blob/main/code-of-conduct.md)
- [X] I searched [issues](https://github.com/search?q=user%3Aremarkjs&type=Issues) and couldn’t find anything (or linked relevant results below)

### Affected packages and versions

unified-lint-rule@2.1.1

### Steps to reproduce

```javascript
const rule = lintRule('foo', (_, file) => {
  file.info("This message should have fatal === null");
});
```

### Expected behavior

The `message.fatal` should remain `null`.

### Actual behavior

The `message.fatal` gets [rewritten by severity](https://github.com/remarkjs/remark-lint/blob/6ca7eeff07bb5c5e6a0daad68520b46e3d56ad8e/packages/unified-lint-rule/lib/index.js#L64). In default cause, set to `false`.

### Runtime

Node v17

### Package manager

yarn 2, Other (please specify in steps to reproduce)

### OS

macOS

**Comment by wooorm** [comment]
That’s its only focus: to set ruleIds and to set a severity.
It’s not too hard to *not* use this utility, and instead set ruleIds and severities yourself.
Particularly, if you are already setting `fatal` manually, why not set the source of messages too, and not use this?

**Comment by ritave** [comment]
I'm not setting `fatal` manually.

The lint rule I've written follows the API set out by remark-lint with severity as option.  
But it also [writes some `file.info()` messages](https://github.com/MetaMask/SIPs/blob/3016e889b34bd85ed58d7faf52081ab4663d95d3/tools/validate/src/rules/git-updated.ts#L33) as well, that aren't a linting problem, but an informative clause to the user.

**Comment by wooorm** [comment]
> I'm not setting fatal manually.

I believe I understand what you mean.
However, `file.message` is the generic way to add messages. `file.info` and `file.fail` wrap that to set `fatal`.

> But it also writes some `file.info()` messages

Thanks for sharing more info.

`unified-lint-rule` is a small wrapper around a unified plugin, which you can also make yourself, which allows:
a) to add metadata common to all messages emitted in a rule. That is to say, it assumes all messages are about the same “problem”. Sometimes the message texts are slightly different (add x spaces or remove x spaces), but it’s all about the spaces being incorrect.
b) users to change, with comments or as options, whether all those messages are warnings (default), errors, or info.

Your example code seems to do several things in one. I think that’s fine as a plugin, but isn’t really how I see lint rules.
Particularly, I don’t see that all messages you emit in your code should receive the same `ruleId` and even perhaps `url`.

Alternatively, I can also imagine that in your code you’d change:
- the message for `--is-shallow-repository`: into an error
- the messages of `The file is not in a git repository` and `File is not committed to git repository`: into debug calls

If not, I think it’s better as a plugin (“sip-git-updated”), where different messages are emitted (such as `sip-git-updated:git-missing`, `sip-git-updated:git-shallow`).

**Comment by ritave** [comment]
That's one possibility.

All those checks are preconditions for actually [running the main rule](https://github.com/MetaMask/SIPs/blob/3016e889b34bd85ed58d7faf52081ab4663d95d3/tools/validate/src/rules/git-updated.ts#L55). Since the validator may run in an environment without git, and that's ok, those preconditions just inform the user that this rule could not be fully checked but otherwise is in acceptable state. It's information that "Hey, maybe your environment is not properly configured so you should check that? But I'm not judging, as far as I'm concerned, this is good to go"

Another way I could do it, is enable / disable the lint rule before adding it as a plugin, just thought doing it inside the rule itself is cleaner.

I'm happy to write my own helper if you think this issue is out of scope for unified-lint-rule.

**Comment by wooorm** [comment]
> Another way I could do it, is enable / disable the lint rule before adding it as a plugin

I do not understand you.

Have you looked at the code of this package? It's tiny for the metadata part. It's giant for configuring 'fatal' by users. Which you don't seem to want.

If you don't want the latter part, you can use a normal plugin, at no cost in line size or complexity!

**Comment by ritave** [comment]
I have rules that are run only under certain preconditions.

For example, I validate properties on front-matter, but I only run those rules if the front-matter is a valid YAML.

For this rule, I want to validate git - front-matter property relationship, but only if (1) user has git and (2) the property exists.

Otherwise the rule doesn't run.

The rule doesn't run if there's no git, but this might be confusing to the user, so I'd like to inform the user that we've tried to run this rule, but we couldn't due to their environment.

---

I read the source code of this package, I'm very impressed by the elegance of the design of the whole remark project.

This issue proposes to extend the contract of this utility -> `file.message()` should be used to record linting problems, which should follow the normal contract of `severity`, while `file.info()` could be used for auxiliary information that [ignore severity altogether](https://github.com/MetaMask/SIPs/blob/main/tools/validate/.yarn/patches/unified-lint-rule-npm-2.1.1-237c9e9f5f.patch) and always have `fatal: null`.

**Comment by ritave** [comment]
Thank you for the changes. The reason I'm avoiding a custom plugin for a rule is because I want to stay consistent with upstream.

I just though this change would benefit anyone else using unified-lint-rule in their own projects as `file.info()` returning errors is a bit unexpected and seemed like a useful case.

I'm very thankful for your patient approach, I'll close this issue as it looks this change is out of scope for this package.

**Comment by wooorm** [comment]
> consistent with upstream

Not sure what upstream is. Most of unified uses plain plugins. There are only a few cases where this utility is useful: those that want users to configure severities and not want to add metadata to each message manually!

> I'm very thankful for your patient approach

<3

> I just though this change would benefit anyone else using `unified-lint-rule` in their own projects as `file.info()`

I can see *some* value in it. But for this case, the alternative (a normal plugin), seems to me like a more than good enough solution!
